### PR TITLE
Make Dropout and BatchNormalization Training-friendly

### PR DESCRIPTION
**Pull request by SherlockNoMad**
As we have discussed in the ONNX training working group, we wish to make ONNX ops' training friendly. 
To start with, we are adding "is_train" as an input to Dropout and BatchNormalization.

- Why as an input, why not attribute?
During training, we also need to performs evaluation periodically to check model's performance. This requires flipping the operation mode on-the-fly. As attribute is usually a constant value for a model, it would be tricky to override it during training. Exposing it as an input allows user to change the value through data feeding from model input. 

- Why 'is_train', why not 'is_test'? 
In the previous versions of Dropout and BatchNormalization, the mode is named as 'is_test'. As ONNX is still mainly for inference purpose, it's better to have the default mode as inference, so that the change won't affect the existing models. By setting is_train = true, we enable the training mode. This is more intuitive than setting is_test = false. 

This PR should also address issues #1042 

@houseroad @pranavsharma @ebarsoum @linkerzhang @prasanthpul @yuanbyu

### CODE REVIEW

**Comment by pranavsharma** [comment]
LGTM 👍

**Comment by Adamsmets** [comment]
One time framing down has ended on the ES. Will the sisters turn on Euro closing? 
#tradingforex #forex #tradeforex #forextrader #forexmarket #stocks #forextrading #bitcointrading #forexprofit #cryptotrading #forexsignals #currencytrading #forexeducation #forextraining

**Comment by hobei** [comment]
Do you see any issue with the 'is_train' input being defined separately on each operation, which could potentially result in some operations in training mode and some in testing mode in the same model?

**Comment by SherlockNoMad** [comment]
@hobei 
Hi Simon, I actually see this as an advantage to have separated control on each op. 

Imagine you have a model with two dropout node, one locates in the first few layer, one in the last few layer. If you wish to only fine tune the last few layers, one can set the second dropout node to train mode while keep the first dropout in the test mode.

**Comment by hobei** [comment]
@SherlockNoMad 
Hi Sherlock. I agree that this is an potential feature. My concern is how to make it consistent if you would want  apply the same control to operations that do not have the 'is_train' input. How do you identify to fine tune layers which do not contain Dropout and Batchnorm?

**Comment by chinhuang007** [comment]
@SherlockNoMad I think the proposed solution is sound, handling different behaviors of certain operators between training and inference modes. The question is how do we know we have covered all operators that need to have this optional input? Another question is what do we do with a Dropout/BatchNorm with is_train=true while creating/optimizing the inference graph?

**Comment by chinhuang007** [comment]
@hobei While await for Sherlock's response, I just want to provide my view... The feature of "defining the fine tune layers" seems a separate interesting topic. The is_train input is to determine the expected behavior at the individual operator level. To selectively fine tune certain layers, or some sub-graphs, during training, we might need to introduce something new.

**Comment by SherlockNoMad** [comment]
@chinhuang007 
I have gone through the operator list and only identified these two operators behave differently during inference and training. As we encounter more such operators, we will need to add this optional input for them. Hopefully, there are not many of them. 
Inference graph should have is_train set to false, or left empty. This should be handled by the model constructor/converter/optimizer.

**Comment by SherlockNoMad** [comment]
@hobei @chinhuang007 
Identifying the layers to fine tune is indeed a separate topic. In my opinion, it should be handle by the training runtime/backend.

File: onnx/defs/nn/defs.cc [hunk 1]
```diff
@@ -1281,6 +1281,12 @@ ONNX_OPERATOR_SET_SCHEMA(
             "var",
             "running (training) or estimated (testing) variance tensor of shape (C).",
             "T")
+        .Input(
+            5,
+            "is_train",
+            "If set to nonzero, run spatial batch normalization in training mode, default is 0.",
```
Comment (spandantiwari) [inline onnx/defs/nn/defs.cc]: I think is_train is type tensor(boolean). Maybe we should consider using {true, false} instead.

**Comment by hobei** [comment]
@SherlockNoMad  
I have another comment, specifically for batch norm. When the is_test attribute was removed from op set 7 it was replaced (as I understand it) with the number of outputs instead indicating the mode (training/testing).

```
Output case #1: Y, mean, var, saved_mean, saved_var (training mode) Output case #2: Y (test mode)
```

Can you please elaborate on how the value of the new 'is_train' input relates to the number of outputs of batchnorm? As I understand it the number of outputs in a model is fixed, as they may be used as inputs to other operations.

**Comment by SherlockNoMad** [comment]
@hobei , I have updated the doc to further elaborate it.

Output case #1: Y, mean, var, saved_mean, saved_var (training mode)
Output case #2: Y (test mode, where other outputs will not be populated)

The number of outputs is indeed fixed. During testing, the latter 4 outputs are not populated and should not be consumed by other nodes.

File: onnx/defs/nn/defs.cc [hunk 2]
```diff
@@ -1719,11 +1731,20 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+            "seed",
+            "(Optional) Seed to the random generator, if not specified we will auto generate one.",
             AttributeProto::FLOAT,
-            0.5f)
+            OPTIONAL)
         .Input(0, "data", "The input data as Tensor.", "T")
+        .Input(
+            1,
+            "ratio",
+            "The ratio of random dropout, with value in [0, 1]. If this input was not set, "
```
Comment (edgchen1) [inline onnx/defs/nn/defs.cc]: ratio's range should be [0,1)

**Comment by wschin** [comment]
Since #2568 is merged, we can close this PR.
